File: windows/tests/Elevate.Core.Tests/ScopeTreeTests.cs
using Elevate.Core.Models;
using Elevate.Core.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class ScopeTreeTests
{
    private const string SubA = "/subscriptions/aaaaaaaa-0000-0000-0000-000000000000";
    private const string SubB = "/subscriptions/bbbbbbbb-0000-0000-0000-000000000000";
    private const string Mg = "/providers/Microsoft.Management/managementGroups/platform";

    private static EligibleRole Role(string name, string scope, string? detail = null) =>
        new(new RoleKey("i", "t", new AzureResourceScope(scope, "rd-" + name)),
            name, RoleSource.Discovered, RolePolicy.ManualDefault, Detail: detail);

    [Fact]
    public void Build_NestsResourceGroupsUnderTheirSubscription()
    {
        var tree = ScopeTree.Build(
        [
            Role("Contributor", SubA + "/resourceGroups/prod", "prod · resource group"),
            Role("Reader", SubA + "/resourceGroups/test", "test · resource group"),
            Role("Owner", SubA, "Platform · subscription"),
        ]);

        tree.Should().ContainSingle();
        var sub = tree[0];
        sub.Kind.Should().Be(ArmScopeKind.Subscription);
        sub.Title.Should().Be("Platform", because: "the role held on the subscription itself carries its caption");
        sub.Roles.Select(r => r.DisplayName).Should().Equal("Owner");
        sub.Children.Select(c => c.Title).Should().Equal("prod", "test");
        sub.RoleCount.Should().Be(3);
    }

    [Fact]
    public void Build_ShowsTheArmNameWhenNoRoleSitsOnTheScopeItself()
    {
        var tree = ScopeTree.Build(
        [
            Role("Contributor", SubA + "/resourceGroups/prod", "prod · resource group"),
            Role("Reader", SubA + "/resourceGroups/test", "test · resource group"),
        ]);

        // Nothing is eligible on the subscription, so the service never named it.
        tree[0].DisplayName.Should().BeNull();
        tree[0].Title.Should().Be("aaaaaaaa-0000-0000-0000-000000000000");
    }

    [Fact]
    public void Build_PutsManagementGroupsBesideSubscriptionsRatherThanAboveThem()
    {
        // ARM never repeats the management group in a subscription's scope, so the string cannot
        // tell us the subscription sits under it.
        var tree = ScopeTree.Build([Role("Reader", Mg, "Platform · management group"), Role("Owner", SubA, "A · subscription")]);
        tree.Select(n => n.Kind).Should().Equal(ArmScopeKind.ManagementGroup, ArmScopeKind.Subscription);
    }

    [Fact]
    public void Build_IgnoresRolesThatAreNotAzureResourceRoles()
    {
        var entra = new EligibleRole(
            new RoleKey("i", "t", new EntraDirectoryScope("rd", "/")), "Global Reader", RoleSource.Discovered, RolePolicy.ManualDefault);
        ScopeTree.Build([entra, Role("Owner", SubA)]).Should().ContainSingle();
    }

    [Fact]
    public void Build_SortsRootsByKindThenTitle()
    {
        var tree = ScopeTree.Build(
        [
            Role("Owner", SubB, "Zulu · subscription"),
            Role("Owner", SubA, "Alpha · subscription"),
            Role("Reader", Mg, "Platform · management group"),
        ]);
        tree.Select(n => n.Title).Should().Equal("Platform", "Alpha", "Zulu");
    }

    [Fact]
    public void Flatten_ElidesAScopeThatLeadsToOneRoleAndBranchesNowhere()
    {
        // The motivating case: many subscriptions, one eligibility each. A header per subscription
        // would double the rows and show nothing the role row does not already say.
        var tree = ScopeTree.Build([Role("Contributor", SubA, "Alpha · subscription"), Role("Contributor", SubB, "Zulu · subscription")]);
        var rows = ScopeTree.Flatten(tree);

        rows.Should().HaveCount(2);
        rows.Should().OnlyContain(r => !r.IsScope);
        rows.Should().OnlyContain(r => r.Depth == 0);
    }

    [Fact]
    public void Flatten_DrawsAHeaderAsSoonAsAScopeHasSomethingToBranchInto()
    {
        var tree = ScopeTree.Build(
        [
            Role("Owner", SubA, "Alpha · subscription"),
            Role("Contributor", SubA + "/resourceGroups/prod", "prod · resource group"),
        ]);
        var rows = ScopeTree.Flatten(tree);

        rows.Select(r => (r.Depth, r.IsScope, r.Role?.DisplayName)).Should().Equal(
            (0, true, null),
            (1, false, "Owner"),
            (1, false, "Contributor"));
    }

    [Fact]
    public void Flatten_DrawsAHeaderWhenOneScopeHoldsSeveralRoles()
    {
        var tree = ScopeTree.Build([Role("Owner", SubA, "Alpha · subscription"), Role("Reader", SubA, "Alpha · subscription")]);
        var rows = ScopeTree.Flatten(tree);
        rows[0].IsScope.Should().BeTrue();
        rows.Skip(1).Select(r => r.Role!.DisplayName).Should().Equal("Owner", "Reader");
    }

    [Fact]
    public void Flatten_StopsAtACollapsedNode()
    {
        var tree = ScopeTree.Build(
        [
            Role("Owner", SubA, "Alpha · subscription"),
            Role("Contributor", SubA + "/resourceGroups/prod", "prod · resource group"),
        ]);
        var rows = ScopeTree.Flatten(tree, n => n.Scope == SubA);

        rows.Should().ContainSingle();
        rows[0].IsScope.Should().BeTrue();
    }

    [Fact]
    public void Flatten_NestsAResourceUnderItsResourceGroup()
    {
        var rg = SubA + "/resourceGroups/prod";
        var tree = ScopeTree.Build(
        [
            Role("Owner", rg, "prod · resource group"),
            Role("Reader", rg + "/providers/Microsoft.Compute/virtualMachines/web01", "web01 · virtualmachines"),
            Role("Reader", rg + "/providers/Microsoft.Compute/virtualMachines/web02", "web02 · virtualmachines"),
        ]);
        var rows = ScopeTree.Flatten(tree);

        // The subscription is folded away — it holds no role of its own and leads to a single child.
        rows.Select(r => (r.Depth, r.IsScope)).Should().Equal(
            (0, true), (1, false), (1, false), (1, false));
        rows[0].Node!.Kind.Should().Be(ArmScopeKind.ResourceGroup);
    }

    [Fact]
    public void Build_FoldsAPassThroughScopeIntoTheNodeBelowIt()
    {
        var rg = SubA + "/resourceGroups/prod";
        var tree = ScopeTree.Build(
        [
            Role("Owner", rg, "prod · resource group"),
            Role("Reader", rg, "prod · resource group"),
        ]);

        tree.Should().ContainSingle();
        tree[0].Kind.Should().Be(ArmScopeKind.ResourceGroup);
        tree[0].Ancestors.Should().Equal("aaaaaaaa-0000-0000-0000-000000000000");
        tree[0].Path.Should().Be("aaaaaaaa-0000-0000-0000-000000000000 / prod");
        tree[0].Scope.Should().Be(rg, because: "selecting the folded node must still reach the same subtree");
    }

    [Fact]
    public void Build_DoesNotFoldAScopeThatHoldsARoleOfItsOwn()
    {
        var tree = ScopeTree.Build(
        [
            Role("Owner", SubA, "Alpha · subscription"),
            Role("Reader", SubA + "/resourceGroups/prod", "prod · resource group"),
        ]);
        tree[0].Kind.Should().Be(ArmScopeKind.Subscription);
        tree[0].Ancestors.Should().BeEmpty();
    }

    [Fact]
    public void Build_DoesNotFoldAScopeThatBranches()
    {
        var tree = ScopeTree.Build(
        [
            Role("Owner", SubA + "/resourceGroups/prod", "prod · resource group"),
            Role("Reader", SubA + "/resourceGroups/test", "test · resource group"),
        ]);
        tree[0].Kind.Should().Be(ArmScopeKind.Subscription);
        tree[0].Children.Should().HaveCount(2);
    }

    [Fact]
    public void AllRoles_ReachesEverythingBelowTheNode()
    {
        var tree = ScopeTree.Build(
        [
            Role("Owner", SubA, "Alpha · subscription"),
            Role("Contributor", SubA + "/resourceGroups/prod", "prod · resource group"),
            Role("Reader", SubA + "/resourceGroups/prod/providers/Microsoft.Compute/virtualMachines/web01", "web01 · vm"),
        ]);
        tree[0].AllRoles.Select(r => r.DisplayName).Should().BeEquivalentTo("Owner", "Contributor", "Reader");
        tree[0].RoleCount.Should().Be(3);
    }
}

File: macos/Tests/ElevateCoreTests/ScopeTreeTests.swift
import Testing
@testable import ElevateCore

/// Port of the C# `ScopeTreeTests`.
struct ScopeTreeTests {
    let subA = "/subscriptions/aaaaaaaa-0000-0000-0000-000000000000"
    let subB = "/subscriptions/bbbbbbbb-0000-0000-0000-000000000000"
    let mg = "/providers/Microsoft.Management/managementGroups/platform"

    private func role(_ name: String, _ scope: String, detail: String? = nil) -> EligibleRole {
        EligibleRole(
            key: RoleKey(identityId: "i", tenantId: "t", scope: .azureResource(scope: scope, roleDefinitionId: "rd-" + name)),
            displayName: name, detail: detail, source: .discovered, policy: .manualDefault)
    }

    @Test func buildNestsResourceGroupsUnderTheirSubscription() {
        let tree = ScopeTree.build([
            role("Contributor", subA + "/resourceGroups/prod", detail: "prod · resource group"),
            role("Reader", subA + "/resourceGroups/test", detail: "test · resource group"),
            role("Owner", subA, detail: "Platform · subscription"),
        ])

        #expect(tree.count == 1)
        let sub = tree[0]
        #expect(sub.kind == .subscription)
        // The role held on the subscription itself carries its caption.
        #expect(sub.title == "Platform")
        #expect(sub.roles.map(\.displayName) == ["Owner"])
        #expect(sub.children.map(\.title) == ["prod", "test"])
        #expect(sub.roleCount == 3)
    }

    @Test func buildShowsTheArmNameWhenNoRoleSitsOnTheScopeItself() {
        let tree = ScopeTree.build([
            role("Contributor", subA + "/resourceGroups/prod", detail: "prod · resource group"),
            role("Reader", subA + "/resourceGroups/test", detail: "test · resource group"),
        ])

        // Nothing is eligible on the subscription, so the service never named it.
        #expect(tree[0].displayName == nil)
        #expect(tree[0].title == "aaaaaaaa-0000-0000-0000-000000000000")
    }

    @Test func buildPutsManagementGroupsBesideSubscriptionsRatherThanAboveThem() {
        // ARM never repeats the management group in a subscription's scope, so the string cannot
        // tell us the subscription sits under it.
        let tree = ScopeTree.build([
            role("Reader", mg, detail: "Platform · management group"),
            role("Owner", subA, detail: "A · subscription"),
        ])
        #expect(tree.map(\.kind) == [.managementGroup, .subscription])
    }

    @Test func buildIgnoresRolesThatAreNotAzureResourceRoles() {
        let entra = EligibleRole(
            key: RoleKey(identityId: "i", tenantId: "t", scope: .entraDirectory(roleDefinitionId: "rd", directoryScopeId: "/")),
            displayName: "Global Reader", source: .discovered, policy: .manualDefault)
        #expect(ScopeTree.build([entra, role("Owner", subA)]).count == 1)
    }

    @Test func buildSortsRootsByKindThenTitle() {
        let tree = ScopeTree.build([
            role("Owner", subB, detail: "Zulu · subscription"),
            role("Owner", subA, detail: "Alpha · subscription"),
            role("Reader", mg, detail: "Platform · management group"),
        ])
        #expect(tree.map(\.title) == ["Platform", "Alpha", "Zulu"])
    }

    @Test func flattenElidesAScopeThatLeadsToOneRoleAndBranchesNowhere() {
        // The motivating case: many subscriptions, one eligibility each. A header per subscription
        // would double the rows and show nothing the role row does not already say.
        let tree = ScopeTree.build([
            role("Contributor", subA, detail: "Alpha · subscription"),
            role("Contributor", subB, detail: "Zulu · subscription"),
        ])
        let rows = ScopeTree.flatten(tree)

        #expect(rows.count == 2)
        #expect(rows.allSatisfy { !$0.isScope })
        #expect(rows.allSatisfy { $0.depth == 0 })
    }

    @Test func flattenDrawsAHeaderAsSoonAsAScopeHasSomethingToBranchInto() {
        let tree = ScopeTree.build([
            role("Owner", subA, detail: "Alpha · subscription"),
            role("Contributor", subA + "/resourceGroups/prod", detail: "prod · resource group"),
        ])
        let rows = ScopeTree.flatten(tree)

        #expect(rows.map(\.depth) == [0, 1, 1])
        #expect(rows.map(\.isScope) == [true, false, false])
        #expect(rows.compactMap { $0.role?.displayName } == ["Owner", "Contributor"])
    }

    @Test func flattenDrawsAHeaderWhenOneScopeHoldsSeveralRoles() {
        let tree = ScopeTree.build([
            role("Owner", subA, detail: "Alpha · subscription"),
            role("Reader", subA, detail: "Alpha · subscription"),
        ])
        let rows = ScopeTree.flatten(tree)
        #expect(rows[0].isScope)
        #expect(rows.dropFirst().compactMap { $0.role?.displayName } == ["Owner", "Reader"])
    }

    @Test func flattenStopsAtACollapsedNode() {
        let tree = ScopeTree.build([
            role("Owner", subA, detail: "Alpha · subscription"),
            role("Contributor", subA + "/resourceGroups/prod", detail: "prod · resource group"),
        ])
        let rows = ScopeTree.flatten(tree) { $0.scope == subA }

        #expect(rows.count == 1)
        #expect(rows[0].isScope)
    }

    @Test func flattenNestsAResourceUnderItsResourceGroup() {
        let rg = subA + "/resourceGroups/prod"
        let tree = ScopeTree.build([
            role("Owner", rg, detail: "prod · resource group"),
            role("Reader", rg + "/providers/Microsoft.Compute/virtualMachines/web01", detail: "web01 · virtualmachines"),
            role("Reader", rg + "/providers/Microsoft.Compute/virtualMachines/web02", detail: "web02 · virtualmachines"),
        ])
        let rows = ScopeTree.flatten(tree)

        // The subscription is folded away — it holds no role of its own and leads to a single child.
        #expect(rows.map(\.depth) == [0, 1, 1, 1])
        #expect(rows.map(\.isScope) == [true, false, false, false])
        #expect(rows[0].node.kind == .resourceGroup)
    }

    @Test func buildFoldsAPassThroughScopeIntoTheNodeBelowIt() {
        let rg = subA + "/resourceGroups/prod"
        let tree = ScopeTree.build([
            role("Owner", rg, detail: "prod · resource group"),
            role("Reader", rg, detail: "prod · resource group"),
        ])

        #expect(tree.count == 1)
        #expect(tree[0].kind == .resourceGroup)
        #expect(tree[0].ancestors == ["aaaaaaaa-0000-0000-0000-000000000000"])
        #expect(tree[0].path == "aaaaaaaa-0000-0000-0000-000000000000 / prod")
        // Selecting the folded node must still reach the same subtree.
        #expect(tree[0].scope == rg)
    }

    @Test func buildDoesNotFoldAScopeThatHoldsARoleOfItsOwn() {
        let tree = ScopeTree.build([
            role("Owner", subA, detail: "Alpha · subscription"),
            role("Reader", subA + "/resourceGroups/prod", detail: "prod · resource group"),
        ])
        #expect(tree[0].kind == .subscription)
        #expect(tree[0].ancestors.isEmpty)
    }

    @Test func buildDoesNotFoldAScopeThatBranches() {
        let tree = ScopeTree.build([
            role("Owner", subA + "/resourceGroups/prod", detail: "prod · resource group"),
            role("Reader", subA + "/resourceGroups/test", detail: "test · resource group"),
        ])
        #expect(tree[0].kind == .subscription)
        #expect(tree[0].children.count == 2)
    }

    @Test func allRolesReachesEverythingBelowTheNode() {
        let tree = ScopeTree.build([
            role("Owner", subA, detail: "Alpha · subscription"),
            role("Contributor", subA + "/resourceGroups/prod", detail: "prod · resource group"),
            role("Reader", subA + "/resourceGroups/prod/providers/Microsoft.Compute/virtualMachines/web01", detail: "web01 · vm"),
        ])
        #expect(Set(tree[0].allRoles.map(\.displayName)) == ["Owner", "Contributor", "Reader"])
        #expect(tree[0].roleCount == 3)
    }
}
